### PR TITLE
 Avoid C-style cast in Trilinos-related files

### PR DESCRIPTION
**Pull request by masterleinad**
Part of #7459.
This PR replaces all C-style casts in all `Trilinos`-related headers and source files.
In a lot of places, we need to use `static_cast`s instead of `dynamic_cast`s for the code to work. This is clearly suspicious. As for the `PETSc`-related code, we are using a lot of `reinterpret_cast`s. This also appears to be fragile.
This PR already incorporates the relevant changes form #7494 and removes `static_cast`s where they are not necessary to improve readability.

### CODE REVIEW

File: source/lac/trilinos_precondition.cc [hunk 1]
```diff
@@ -173,7 +173,7 @@ namespace TrilinosWrappers
                       additional_data.overlap));
 
     Ifpack_Preconditioner *ifpack =
-      static_cast<Ifpack_Preconditioner *>(preconditioner.get());
+      dynamic_cast<Ifpack_Preconditioner *>(preconditioner.get());
```
Comment (drwells) [inline source/lac/trilinos_precondition.cc]: Nice catch!

File: include/deal.II/lac/trilinos_vector.h [hunk 2]
```diff
@@ -1613,7 +1613,8 @@ namespace TrilinosWrappers
             {
               const int ierr = vector->ReplaceGlobalValues(
                 1,
-                (const TrilinosWrappers::types::int_type *)(&row),
+                reinterpret_cast<const TrilinosWrappers::types::int_type *>(
```
Comment (bangerth) [inline include/deal.II/lac/trilinos_vector.h]: Doesn't this suggest that we gave `row` above the wrong type?

Comment (masterleinad) [reply include/deal.II/lac/trilinos_vector.h]: Yes, we are relying on `unsigned long long` having the same bit representation as `long long` for the values we are using here.

Comment (bangerth) [reply include/deal.II/lac/trilinos_vector.h]: I think I would prefer if we handle this through assignment -- change the data type above and then you don't have to use the reinterpret_cast here and below.

**Comment by drwells** [comment]
It looks like @bangerth's concern has been addressed so I suggest we merge once he confirms he is happy.

**Comment by masterleinad** [comment]
Is there anything more I should address here?

**Comment by masterleinad** [comment]
@bangerth Any further requests?

**Pull request merged**